### PR TITLE
Pullrequest relative localization no north filter

### PR DESCRIPTION
**Pull request by coppolam**
Added a relative localization filter that does not require the drones to have knowledge of a common heading (such as North). This allows to be independent of errors coming from the magnetometer or the building bias that may be caused by a gyroscope. The filter is as implemented and used in van der Helm et al. "On-board Range-based Relative Localization for Micro Aerial Vehicles in indoor Leader-Follower Flight". https://arxiv.org/abs/1805.07171.

### CODE REVIEW

File: sw/airborne/math/pprz_algebra.h [hunk 1]
```diff
@@ -42,10 +42,13 @@ extern "C" {
 
 #include <float.h>  /* for FLT_EPSILON */
 #include <string.h> /* for memcpy      */
-#include "std.h" /* for ABS */
+#include "std.h"  /* for ABS */
+#include "math.h" /* for log and fabs */
 
 #define SQUARE(_a) ((_a)*(_a))
 
+#define MAXIMUM_VALUE(_a,_b) (_a) < (_b) ? (_b) : (_a)
```
Comment (kirkscheper) [inline sw/airborne/math/pprz_algebra.h]: Already defined here https://github.com/paparazzi/paparazzi/blob/master/sw/include/std.h#L95

File: sw/airborne/math/pprz_algebra_float.h [hunk 2]
```diff
@@ -101,6 +101,19 @@ struct FloatRates {
...
 
+/*
+ * Returns the log of v in base of n
+ */
+static inline float float_log_n(float v, float n)
+{
+  if (v == 0.0) {
+    return - 1.0E+30;
+  } else {
+    return log(fabs(v)) / log(n);
```
Comment (kirkscheper) [inline sw/airborne/math/pprz_algebra_float.h]: logf

File: sw/airborne/math/pprz_algebra_float.c [hunk 3]
```diff
@@ -873,3 +873,89 @@ void float_mat_invert(float **o, float **mat, int n)
...
+float float_mat_norm_li(float **a, int m, int n)
+{
+  float row_sum;
+  float value;
+
+  value = 0.0;
+  for (int i = 0; i < m; i++) {
+    row_sum = 0.0;
+    for (int j = 0; j < n; j++) {
+      row_sum = row_sum + fabs(a[i][j]);
```
Comment (kirkscheper) [inline sw/airborne/math/pprz_algebra_float.c]: fabsf

File: sw/airborne/math/pprz_algebra_float.h [hunk 4]
```diff
@@ -721,6 +767,58 @@ static inline void float_mat_vect_mul(float *o, float **a, float *b, int m, int
...
+ *
+ * a: [m x n]
+ * k: [1 x 1]
+ */
+static inline void float_mat_scale(float **a, float k, int m, int n)
+{
+  int i, j;
+  for (i = 0; i < m; i++) {
+    for (j = 0; j < n; j++) {
+      if (a[i][j] != 0.0) {
```
Comment (kirkscheper) [inline sw/airborne/math/pprz_algebra_float.h]: This inequality is not a good idea, as float doesn't represent 0 well. You usually check if it is larger than some epsilon. https://bitbashing.io/comparing-floats.html

 Also, why do you do this check, the next line of computation is fine if a is 0.

File: sw/airborne/math/pprz_algebra_float.h [hunk 5]
```diff
@@ -721,6 +767,58 @@ static inline void float_mat_vect_mul(float *o, float **a, float *b, int m, int
...
+ * a: [m x n]
+ * b: [m x n]
+ * k: [1 x 1]
+ */
+static inline void float_mat_sum_scaled(float **a, float **b, float k, int m, int n)
+{
+  int i, j;
+  for (i = 0; i < m; i++) {
+    for (j = 0; j < n; j++) {
+      if (a[i][j] != 0.0) {
```
Comment (kirkscheper) [inline sw/airborne/math/pprz_algebra_float.h]: Same here

File: sw/airborne/math/pprz_algebra_float.h [hunk 6]
```diff
@@ -721,6 +767,58 @@ static inline void float_mat_vect_mul(float *o, float **a, float *b, int m, int
...
+        a[i][j] += k * b[i][j];
+      }
+    }
+  }
+}
+
+
+/** a += k*b, where k is a scalar value
```
Comment (kirkscheper) [inline sw/airborne/math/pprz_algebra_float.h]: The description here is the same as above. The code looks the same too...

File: sw/airborne/modules/relative_localization_filter/discrete_ekf_no_north.c [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+/**
+ * @file "modules/relativelocalizationfilter/discrete_ekf_no_north.c"
+ * @author Steven van der Helm, Mario Coppola
+ * Discrete Extended Kalman Filter for Relative Localization without requiring knowledge of North
+ * This implementation is from
+ * van der Helm et al. "On-board Range-based Relative Localization for Micro Aerial Vehicles in indoor Leader-Follower Flight." (2018).
+ * Available at https://arxiv.org/pdf/1805.07171.pdf
+ */
+
+#include "discrete_ekf_no_north.h"
```
Comment (kirkscheper) [inline sw/airborne/modules/relative_localization_filter/discrete_ekf_no_north.c]: This is usually written out in long for i.e. modules/relative_localization_filter/discrete_ekf_no_north.h

File: sw/airborne/modules/relative_localization_filter/discrete_ekf.h [hunk 8]
```diff
@@ -18,7 +18,7 @@
  * <http://www.gnu.org/licenses/>.
  */
 /**
- * @file "modules/relativelocalizationfilter/discreteekf.h"
+ * @file "modules/relativelocalizationfilter/discrete_ekf.h"
```
Comment (kirkscheper) [inline sw/airborne/modules/relative_localization_filter/discrete_ekf.h]: Shouldn't this be modules/relative_localization_filter/discrete_ekf.h

File: sw/airborne/modules/relative_localization_filter/discrete_ekf_no_north.c [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with paparazzi; see the file COPYING.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+/**
+ * @file "modules/relativelocalizationfilter/discrete_ekf_no_north.c"
```
Comment (kirkscheper) [inline sw/airborne/modules/relative_localization_filter/discrete_ekf_no_north.c]: modules/relative_localization_filter/discrete_ekf_no_north.c

File: sw/airborne/modules/relative_localization_filter/discrete_ekf.c [hunk 10]
```diff
@@ -18,7 +18,7 @@
  * <http://www.gnu.org/licenses/>.
  */
 /**
- * @file "modules/relativelocalizationfilter/discreteekf.c"
+ * @file "modules/relativelocalizationfilter/discrete_ekf.c"
```
Comment (kirkscheper) [inline sw/airborne/modules/relative_localization_filter/discrete_ekf.c]: modules/relative_localization_filter/discrete_ekf.c

File: sw/airborne/modules/relative_localization_filter/discrete_ekf_no_north.c [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+enum ekf_input {u1dm, v1dm, u2dm, v2dm, r1m, r2m};
+
+void extractPhiGamma(float **inmat, float **phi, float **gamma, int m, int n_a, int n_b)
+{
+  int totalsize = m + n_b;
+  for (int i = 0; i < totalsize; i++) {
+    for (int j = 0; j < totalsize; j++) {
+      if (i < m && j < n_a) {
+        phi[i][j] = inmat[i][j];
+      } else if (i < m && j >= n_a) {
```
Comment (kirkscheper) [inline sw/airborne/modules/relative_localization_filter/discrete_ekf_no_north.c]: This is equivalent to just "else if(i < m)". But more to the point, if you only want points with i < m why not just do for (int i = 0; i < m; i++) and not do these checks...

File: sw/airborne/modules/relative_localization_filter/discrete_ekf_no_north.c [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+ *                                 0[m][n_a] 0[m][n_b] ]
+ */
+void float_mat_combine(float **a, float **b, float **o, int m, int n_a, int n_b)
+{
+  int totalsize = m + n_b;
+  for (int i = 0; i < totalsize; i++) {
+    for (int j = 0; j < totalsize; j++) {
+      if (i < m && j < n_a) {
+        o[i][j] = a[i][j];
+      } else if (i < m && j >= n_a) {
```
Comment (kirkscheper) [inline sw/airborne/modules/relative_localization_filter/discrete_ekf_no_north.c]: same here "else if(i < m)"

File: sw/airborne/modules/relative_localization_filter/discrete_ekf_no_north.c [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+}
+
+/*
+ * Continuous time state transition equation
+ * state is: {x_rel,y_rel,h1,h2,u1,v1,u2,v2,gamma}
+ */
+
+void discrete_ekf_no_north_fsym(float *statein, float *input, float *output)
+{
+  output[0] =  input[r1m] * statein[y12] - statein[u1] + statein[u2] * cos(statein[gam]) - statein[v2] * sin(
```
Comment (kirkscheper) [inline sw/airborne/modules/relative_localization_filter/discrete_ekf_no_north.c]: cosf and sinf

File: sw/airborne/modules/relative_localization_filter/discrete_ekf_no_north.c [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+  output[7] = input[v2dm] - input[r2m] * statein[u2];
+  output[8] = input[r2m] - input[r1m];
+}
+
+/*
+ * Measurement equation, measures range, height1, height2, u1, v1, u2, v2
+ */
+void discrete_ekf_no_north_hsym(float *statein, float *output)
+{
+  output[0] = pow(pow((statein[z1] - statein[z2]), 2.0) + pow(statein[x12], 2.0) + pow(statein[y12], 2.0), 0.5);
```
Comment (kirkscheper) [inline sw/airborne/modules/relative_localization_filter/discrete_ekf_no_north.c]: powf

File: sw/airborne/modules/relative_localization_filter/discrete_ekf_no_north.c [hunk 15]
```diff
@@ -0,0 +1,302 @@
...
+  output[7][4] = -statein[u2];
+  output[8][4] = -1;
+  output[8][5] = 1;
+}
+
+void discrete_ekf_no_north_Hx(float *statein, float **output)
+{
+  MAKE_MATRIX_PTR(_output, output, EKF_N);
+  float_mat_zero(_output, EKF_M, EKF_N);
+  output[0][0] = statein[x12] / (pow(pow(statein[z1] - statein[z2], 2.0) + pow(statein[x12], 2.0) + pow(statein[y12],
```
Comment (kirkscheper) [inline sw/airborne/modules/relative_localization_filter/discrete_ekf_no_north.c]: powf, this happens lots of times...

File: sw/airborne/modules/relative_localization_filter/discrete_ekf_no_north.h [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with paparazzi; see the file COPYING.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+/**
+ * @file "modules/relativelocalizationfilter/discrete_ekf_no_north.h"
```
Comment (kirkscheper) [inline sw/airborne/modules/relative_localization_filter/discrete_ekf_no_north.h]: correct file name here

File: sw/airborne/modules/relative_localization_filter/relative_localization_filter.c [hunk 17]
```diff
@@ -37,40 +37,76 @@
...
 
 #ifndef RL_NUAVS
 #define RL_NUAVS 4 // Maximum expected number of other UAVs
 #endif
 
+/*
+ * NO_NORTH = 1 : The filter runs without a heading reference.
+ * NO_NORTH = 0 : The filter runs while using a shared reference heading.
+ */
+#ifndef NO_NORTH
```
Comment (kirkscheper) [inline sw/airborne/modules/relative_localization_filter/relative_localization_filter.c]: This is not a very descriptive setting name, can you add something with the module name, recommendation RELATIVE_LOCALIZATION_NO_NORTH or RELATIVE_LOCALIZATION_FILTER_NO_NORTH or RL_NO_NORHT like the other settings.

File: sw/airborne/modules/relative_localization_filter/relative_localization_filter.c [hunk 18]
```diff
@@ -37,40 +37,76 @@
...
+#include "discrete_ekf_no_north.h"
+#else
+#include "discrete_ekf.h"
+#endif
+
 int32_t id_array[RL_NUAVS]; // array of UWB IDs of all drones
 uint32_t latest_update_time[RL_NUAVS];
 uint8_t number_filters; // the number of filters running in parallel
-struct discrete_ekf ekf_rl[RL_NUAVS]; 
+#if NO_NORTH
```
Comment (kirkscheper) [inline sw/airborne/modules/relative_localization_filter/relative_localization_filter.c]: can you combine this with the include above on line 53?

File: sw/airborne/modules/relative_localization_filter/discrete_ekf_no_north.c [hunk 19]
```diff
@@ -0,0 +1,302 @@
...
+  output[6] = statein[v2];
+}
+
+void discrete_ekf_no_north_Fx(float *statein, float *input, float **output)
+{
+  MAKE_MATRIX_PTR(_output, output, EKF_N);
+  float_mat_zero(_output, EKF_N, EKF_N);
+  output[0][1] = input[r1m];
+  output[0][4] = -1;
+  output[0][6] = cos(statein[gam]);
```
Comment (kirkscheper) [inline sw/airborne/modules/relative_localization_filter/discrete_ekf_no_north.c]: cosf and sinf everywhere

File: sw/airborne/math/pprz_algebra_float.c [hunk 20]
```diff
@@ -873,3 +873,89 @@ void float_mat_invert(float **o, float **mat, int n)
...
+  MAKE_MATRIX_PTR(_o,  o,  n);
+  MAKE_MATRIX_PTR(_d,  d,  n);
+  MAKE_MATRIX_PTR(_x,  x,  n);
+  MAKE_MATRIX_PTR(_a_copy, a_copy, n);
+
+  float_mat_copy(_a_copy, _a, n, n); // Make a copy of a to compute on
+  a_norm = float_mat_norm_li(_a_copy, n, n);  // Compute the infinity norm of the matrix
+  ee = (int)(float_log_n(a_norm, 2)) + 1;
+  s = MAXIMUM_VALUE(0, ee + 1);
+  t = 1.0 / pow(2.0, s);
```
Comment (kirkscheper) [inline sw/airborne/math/pprz_algebra_float.c]: powf

File: conf/modules/relative_localization_filter.xml [hunk 21]
```diff
@@ -8,10 +8,17 @@
   </doc>
   <header>
     <file name="relative_localization_filter.h"/>
+<!--     <file name="discrete_ekf.h"/>
```
Comment (kirkscheper) [inline conf/modules/relative_localization_filter.xml]: remove

File: sw/airborne/math/pprz_algebra_float.h (hunk 5 above)
Comment (coppolam) [reply sw/airborne/math/pprz_algebra_float.h]: This function is not used actually. Didn't notice it ended up in the pr, I'll get rid of it.

File: sw/airborne/math/pprz_algebra_float.h (hunk 6 above)
Comment (coppolam) [reply sw/airborne/math/pprz_algebra_float.h]: See answer above.

File: sw/airborne/modules/relative_localization_filter/discrete_ekf_no_north.c (hunk 15 above)
Comment (coppolam) [reply sw/airborne/modules/relative_localization_filter/discrete_ekf_no_north.c]: Lesson learned :+1:

File: sw/airborne/math/pprz_algebra_float.h [hunk 22]
```diff
@@ -101,6 +101,19 @@ struct FloatRates {
     while (_a < -M_PI) _a += (2.*M_PI);             \
   }
 
+/*
+ * Returns the log of v in base of n
+ */
+static inline float float_log_n(float v, float n)
+{
+  if (v == 0.0) {
```
Comment (kirkscheper) [inline sw/airborne/math/pprz_algebra_float.h]: Maybe also check n here.

File: conf/modules/relative_localization_filter.xml [hunk 23]
```diff
@@ -13,5 +13,10 @@
   <makefile>
     <file name="relative_localization_filter.c"/>
     <file name="discrete_ekf.c"/>
+    <file name="discrete_ekf_no_north.c"/>
+    <configure name="EKF_XZERO" default="1"/>
+    <configure name="EKF_YZERO" default="1"/>
+    <define name="EKF_XZERO" value="$(EKF_XZERO)"/>
+    <define name="EKF_YZERO" value="$(EKF_YZERO)"/>
```
Comment (kirkscheper) [inline conf/modules/relative_localization_filter.xml]: This looks really weird, @gautierhattenberger can you recommend update here?

File: sw/airborne/math/pprz_algebra_float.h (hunk 22 above)
Comment (coppolam) [reply sw/airborne/math/pprz_algebra_float.h]: I'm not sure what you are referring to here

Comment (kirkscheper) [reply sw/airborne/math/pprz_algebra_float.h]: You check v == 0 but not n == 0

Comment (kirkscheper) [reply sw/airborne/math/pprz_algebra_float.h]: But also, float == 0 is not a good idea... See my previous comments

File: sw/airborne/math/pprz_algebra_float.h [hunk 24]
```diff
@@ -102,13 +102,17 @@ struct FloatRates {
...
 {
-  if (v == 0.0) {
+  if (fabsf(v) < 1e-4) { // avoid inf
     return - 1.0E+30;
-  } else {
+  }
+  else if (fabsf(n) < 1e-4) { // avoid nan
+    return 0;
+  }
+  else {
```
Comment (kirkscheper) [inline sw/airborne/math/pprz_algebra_float.h]: This doesn't look well formatted, can you run fix code style? Also, you don't need the else if or else here as it is implicit with the return statements.

**Pull request merged**